src/org/zaproxy/zap/extension/frontendscanner/ExtensionFrontEndScanner.java

```diff
@@ -0,0 +1,380 @@
+
+        this.extensionScript.registerScriptType(activeScriptType);
+        this.extensionScript.registerScriptType(passiveScriptType);
+    }
+
+    @Override
+    public void postInit() {
+        super.postInit();
+
+        registerUserScripts(activeScriptType);
+        registerUserScripts(passiveScriptType);
+    }
+
+    @Override
+    public boolean canUnload() {
+        // The extension can be dynamically unloaded, all resources used/added can be freed/removed from core.
+        return true;
+    }
+
+    @Override
+    public void unload() {
+        super.unload();
+
+        this.extensionScript.removeScripType(activeScriptType);
+        this.extensionScript.removeScripType(passiveScriptType);
+    }
+
+    private ZapMenuItem getFrontEndScannerMenuToggle() {
+        if (menuFrontEndScanner == null) {
+            menuFrontEndScanner = new ZapMenuItem(PREFIX + ".topmenu.tools.title");
+
+            menuFrontEndScanner.addActionListener(new java.awt.event.ActionListener() {
+                @Override
+                public void actionPerformed(java.awt.event.ActionEvent ae) {
+                    frontEndScannerEnabled = !frontEndScannerEnabled;
+
+                    String value = frontEndScannerEnabled ? "on" : "off";
+                    View.getSingleton().showMessageDialog(
+                        Constant.messages
+                            .getString(ExtensionFrontEndScanner.PREFIX + ".topmenu.tools.msg", value));
+                }
+            });
+        }
+        return menuFrontEndScanner;
+    }
+
+    @Override
+    public String getAuthor() {
+        return Constant.ZAP_TEAM;
+    }
+
+    @Override
+    public String getDescription() {
+        return Constant.messages.getString(PREFIX + ".desc");
+    }
+
+    @Override
+    public URL getURL() {
+        try {
+            return new URL(Constant.ZAP_EXTENSIONS_PAGE);
+        } catch (MalformedURLException e) {
+            return null;
+        }
+    }
+
+    @Override
+    public boolean onHttpRequestSend(HttpMessage msg) {
+        return true;
+    }
+
+    @Override
+    public boolean onHttpResponseReceive(HttpMessage msg) {
+        if (frontEndScannerEnabled && msg.getResponseHeader().isHtml()) {
+            try {
+                String html = msg.getResponseBody().toString();
+
+                Source document = new Source(html);
+                List<Element> heads = document.getAllElements("head");
+                Element head = heads.isEmpty() ? null : heads.get(0);
+
+                if (head != null && msg.getHistoryRef() != null) {
+                    String host = msg.getRequestHeader().getHeader("host");
+                    String frontEndApiUrl = API
+                        .getInstance()
+                        .getCallBackUrl(this.api, "https://" + host);
+
+                    int historyReferenceId = msg.getHistoryRef().getHistoryId();
+
+                    StringBuilder injectedContentBuilder = new StringBuilder()

```

Comment:
This could be initialised with a higher capacity (500?).

Reply:
As for now:
```bash
$ wc -l files/frontendscanner/front-end-scanner.js 
2233 files/frontendscanner/front-end-scanner.js
$ wc -l files/scripts/scripts/client-side-passive/scan-jwt-tokens.js 
114 files/scripts/scripts/client-side-passive/scan-jwt-tokens.js
```

So I will make an initial capacity of 2500.

Reply:
Argh, thanks.

Reply:
```bash
$ wc -c files/scripts/scripts/client-side-passive/scan-jwt-tokens.js
3243 files/scripts/scripts/client-side-passive/scan-jwt-tokens.js
$ wc -c files/frontendscanner/front-end-scanner.js
62858 files/frontendscanner/front-end-scanner.js
```

Ouch. Let's go for 67000 then...

Would be nice to add some minification as part of (or at least depending on) #4883...